drivers/mipi_dbi/mipi_dbi_smartbond.c

```diff
@@ -0,0 +1,610 @@
+			LOG_ERR("Could not read data from SPI");
+			goto _mipi_dbi_read_exit;
+		}
+	}
+
+_mipi_dbi_read_exit:
+
+	/* Restore #CS line */
+	da1469x_lcdc_force_cs_line(false, mipi_dbi_cfg.cs_active_high);
+
+	/* Make sure default LCDC pins are applied upon exit */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("Could not apply MIPI DBI pins' default state (%d)", ret);
+	}
+
+	k_sem_give(&data->device_sem);
+
+	return ret;
+}
+#endif
+
+static int mipi_dbi_smartbond_command_write(const struct device *dev,
+				const struct mipi_dbi_config *dbi_config,
+				uint8_t cmd, const uint8_t *data_buf,
+				size_t len)
+{
+	struct mipi_dbi_smartbond_data *data = dev->data;
+	int ret;
+	lcdc_smartbond_mipi_dbi_cfg mipi_dbi_cfg;
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	mipi_dbi_pm_get(dev);
+
+	/*
+	 * Add an arbitrary valid color format to satisfy subroutine. The MIPI DBI command/data
+	 * engine should not be affected.
+	 */
+	lcdc_smartbond_mipi_dbi_translation(dbi_config, &mipi_dbi_cfg, PIXEL_FORMAT_RGB_565);
+	ret = da1469x_lcdc_mipi_dbi_interface_configure(&mipi_dbi_cfg);
+	if (ret < 0) {
+		k_sem_give(&data->device_sem);
+		return ret;
+	}
+
+	/* Command and accompanied data should be transmitted via the DBIB interface */
+	da1469x_lcdc_send_cmd_data(true, &cmd, 1);
+
+	if (len) {
+		/* Data should be transmitted via the DBIB interface */
+		da1469x_lcdc_send_cmd_data(false, data_buf, len);
+	}
+
+	mipi_dbi_pm_put(dev);
+
+	k_sem_give(&data->device_sem);
+
+	return 0;
+}
+
+static int mipi_dbi_smartbond_write_display(const struct device *dev,
+					const struct mipi_dbi_config *dbi_config,
+					const uint8_t *framebuf,
+					struct display_buffer_descriptor *desc,
+					enum display_pixel_format pixfmt)
+{
+	struct mipi_dbi_smartbond_data *data = dev->data;
+	const struct mipi_dbi_smartbond_config *config = dev->config;
+	lcdc_smartbond_layer_cfg *layer = &data->layer;
+	int ret = 0;
+	lcdc_smartbond_mipi_dbi_cfg mipi_dbi_cfg;
+	uint8_t layer_color = lcdc_smartbond_pixel_to_lcm(pixfmt);
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	mipi_dbi_pm_get(dev);
+
+	/*
+	 * Mainly check if the frame generator is busy with a pending frame update (might happen
+	 * when two frame updates take place one after the other and the display interface is
+	 * quite slow). VSYNC interrupt line should be asserted when the last line is being outputed.
+	 */
+	if (da1469x_lcdc_is_busy()) {
+		LOG_WRN("MIPI DBI host is busy");
+		ret = -EBUSY;
+		goto _mipi_dbi_write_exit;
+	}
+
+	lcdc_smartbond_mipi_dbi_translation(dbi_config, &mipi_dbi_cfg, pixfmt);
+	ret = da1469x_lcdc_mipi_dbi_interface_configure(&mipi_dbi_cfg);
+	if (ret < 0) {
+		goto _mipi_dbi_write_exit;
+	}
+
+	ret = da1469x_lcdc_timings_configure(desc->width, desc->height,
+			(lcdc_smartbond_timing_cfg *)&config->timing_cfg);
+	if (ret < 0) {
+		goto _mipi_dbi_write_exit;
+	}
+
+	LCDC_SMARTBOND_LAYER_CONFIG(layer, framebuf, 0, 0, desc->width, desc->height,
```

Comment:
You should validate that the `buf_size` field matches the `height` and `width` of the descriptor here, because the MIPI DBI API specifies that `buf_size` is the amount of data the controller will send , not `height` and `width`. I chose this distinction because most MIPI DBI controllers likely only care about how many bytes to stream out, which can be determined using `buf_size`.

*Note- the `buf_size` restriction isn't currently documented well. #68378 should clarify this*

Reply:
Understood. In contrast to the majority of MIPI DBI controllers, our LCDC IP requires the `height` and `width` descriptor elements which will implicitly determine the number of bytes to be sent.

Reply:
Sure, I follow that- to be clear, what I am suggesting here is that we add a check like the following:

```
if (desc->width * desc->height * (DISPLAY_BITS_PER_PIXEL(pixfmt) / 8) != desc->buf_size) {
    LOG_ERR("Incorrect buffer size for given width and height");
    return -EINVAL;
}
```

This way a user can't pass in a display buffer descriptor that is otherwise invalid in the MIPI DBI API, and have it work with this driver but not with other implementations

Reply:
Clear. I have fixed it at the beginning of the API.